Add retry when snapshotting configuration.

michelle192837 opened the pull request:
Adding a quick retry to mitigate issues where the configuration is (temporarily) not available on startup.

Added a unit test for this as well (probably could be cleaner, but it should verify that the retry actually works).

ETA: The actual change to add a retry is in config_snapshot.go. Everything else is an update to add a unit test for the retry, and enhance the storage fake to allow that (with a lock to prevent data races on map reads/writes for the fake Opener).

michelle192837 commented:
/retest

michelle192837 commented:
Ah, there's a legitimate race in the new fake logic I believe. Taking a look at that.

michelle192837 commented:
Test should be fixed, ready for review!

chases2 commented on config/snapshot/config_snapshot_test.go:
```diff
@@ -108,6 +108,96 @@ func TestObserve_OnInit(t *testing.T) {
+	tests := []struct {
+		name             string
+		config           *configpb.Configuration
+		configGeneration int64
```
I don't think "configGeneration" is actually a parameter of this test. Consider just hard-coding it when constructing your fake.

michelle192837 replied:
Done, thanks!

chases2 commented on config/snapshot/config_snapshot_test.go:
```diff
@@ -108,6 +108,96 @@ func TestObserve_OnInit(t *testing.T) {
+				},
+			},
+			configGeneration: 1,
+			expectInitial: &configpb.Dashboard{
```
Do we expect an initial config in this test case? It looks like we'd expect just an error. "openOnRetry" is false, so it's just going to keep failing, like if the file just isn't there.

This _happens_ to succeed because your test code returns early after "expectError" and never actually checks this. It's fine to return early so we don't wait 30 seconds for a config we don't expect, but please remove it from the test setup array here.

chases2 commented on config/snapshot/config_snapshot_test.go:
```diff
@@ -108,6 +108,96 @@ func TestObserve_OnInit(t *testing.T) {
+
+			snaps, err := Observe(ctx, nil, client, *path, nil)
+
+			if err != nil {
```
nit: You want to fail if these don't match, right? So you'll want to either check both cases (including "we expected an error, but didn't get one), or a more complex check like `if (err != nil) != test.expectError`

michelle192837 replied:
Ah yup, fixed it up to check both cases. (iirc we do have unit tests in other components that do this if/else style when expecting errors.)

chases2 reviewed (commented):
The code itself looks good (after digging around a bit in this new "retry" library) and the tests demonstrates the API well (ie: you will get a config OR you will get an error, etc), but I think it could be cleaner. Please let me know what you think!

michelle192837 commented:
> The code itself looks good (after digging around a bit in this new "retry" library) and the tests demonstrates the API well (ie: you will get a config OR you will get an error, etc), but I think it could be cleaner. Please let me know what you think!

Thanks for catching these! Agree on the suggestions, implemented them and updated the logic in the fake to be clearer (pretty sure I hit some issues on the boolean conditions).

chases2 reviewed (approved):
Looks great! Thanks for taking the extra time on the tests. :D

Feel free to merge when ready; I only see one tiny nit after the refactor.

/hold

michelle192837 commented on config/snapshot/config_snapshot_test.go:
```diff
@@ -108,6 +108,87 @@ func TestObserve_OnInit(t *testing.T) {
+					},
+				},
+			},
+			expectInitial: &configpb.Dashboard{
```
Done, thanks! o7
